src/Core/Resolvers/SqlMutationEngine.cs

```diff
@@ -535,11 +532,24 @@ public class SqlMutationEngine : IMutationEngine
                                 }
                                 else
                                 {
-                                    // This code block is reached when Update or UpdateIncremental operation does not successfully find the record to
-                                    // update. An exception is thrown which will be returned as a 404 NotFound response.
-                                    throw new DataApiBuilderException(message: "No Update could be performed, record not found",
-                                                                        statusCode: HttpStatusCode.NotFound,
-                                                                        subStatusCode: DataApiBuilderException.SubStatusCodes.UnexpectedRestUpdateOperationFailure);
+                                    if (mutationResultRow is null)
+                                    {
+                                        // Ideally this case should not happen, however may occur due to unexpected reasons,
+                                        // like the DbDataReader being null. We throw an exception
+                                        // which will be returned as an UnexpectedError  
+                                        throw new DataApiBuilderException(message: "An unexpected error occurred while trying to execute the query.",
+                                                                            statusCode: HttpStatusCode.NotFound,
+                                                                            subStatusCode: DataApiBuilderException.SubStatusCodes.UnexpectedError);
+                                    }
+
+                                    if (mutationResultRow.Columns.Count == 0)
+                                    {
+                                        // This code block is reached when Update or UpdateIncremental operation does not successfully find the record to
+                                        // update. An exception is thrown which will be returned as a 404 NotFound response.
+                                        throw new DataApiBuilderException(message: "No Update could be performed, record not found",
+                                                                            statusCode: HttpStatusCode.NotFound,
+                                                                            subStatusCode: DataApiBuilderException.SubStatusCodes.EntityNotFound);
```

Comment:
Yes, I was thinking to do it in a separate PR but added those in this PR (there is an overlap in the changeset, so thought it would be better to do directly in this PR).

Added 2 new substatus codes - `InvalidPrimaryKeyField`, `ItemNotFound`.

`ItemNotFound` - Returned when the item identified by the PK specified does not exist in the database (but the entity name used in the request is valid)
`InvalidPrimaryKeyField` - Returned when the PK field used in the request is invalid (does not exist in the entity's definition).

In both the cases, the status code returned is 404 Not Found( no change in that )

Reply:
InvalidPrimaryKeyField should be 400 Bad Request because the entity in the URL is valid, but the way the user attempts to fetch a record is incorrect and must be corrected. 404 would be valid when the identifier fields are correct, but no record with the specified id is found. We return 400 when the valuetype of the pkvalue used is wrong as well (int instead of char, for example).

Reply:
Thanks for your input. I've modified the logic to return BadRequest when the PK field specified is invalid